Doc/library/socket.rst

```diff
@@ -1084,6 +1084,39 @@ to sockets.
    Return ``True`` if socket is in blocking mode, ``False`` if in
    non-blocking.
 
+   Blocking mode for a Python socket object means that operations
+   like :meth:`recv` or :meth:`sendall` will block the execution of
+   the current thread until they are complete or aborted with a
+   :exc:`socket.timeout` or :exc:`socket.error` errors.
+
+   Non-blocking mode means that socket operations will raise
+   a :exc:`BlockingIOError` if they cannot complete immediately.
+
+   Internally, if a Python socket object has a positive timeout, its
+   underlying file descriptor (FD) is switched to a non-blocking mode.
```

Comment:
I'm wary of documenting internals here. It could be quite confusing to the reader, and doesn't seem to provide any useful information.

Reply:
Yeah, but my confusion with this thing came from me trying to check if the FD is blocking/nonblocking with `fcntl.fcntl(sock, fcntl.F_GETFL, os.O_NONBLOCK) & os.O_NONBLOCK`, and from there it's was quite a painful experience to realize how socket objects work in Python (probably because I was debugging a failing sendfile test after some changes, but whatever).

I agree that all this information might overwhelm an average Python user who just wants to read the documentation.  On the other hand, `socket` module is usually used by more or less advanced Python users.

That said, I'd be OK to keep this detailed documentation piece in `socketmodule.c` if everyone thinks it's too much to have it here.

Reply:
I think it would be best located in `socketmodule.c` indeed. Not sure where.

Reply:
Let's keep it in `sock_settimeout()`, as it's the central place that triggers blocking/non-blocking modes.